api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchengine/solr/SolrFieldInitializer.java

```diff
@@ -21,52 +25,158 @@
+        }
+    }
+
+    private static boolean isFieldTypeExists(String typeName, SolrClient queryEngine,
+            ConcurrentUpdateSolrClient updateEngine) throws Exception {
+        try {
+            SolrQuery query = new SolrQuery();
+            query.add(CommonParams.QT, "/schema/fieldtypes");
+            QueryResponse response = queryEngine.query(query);
+            ArrayList<SimpleOrderedMap> typeList = (ArrayList<SimpleOrderedMap>) response.getResponse()
+                    .get("fieldTypes");
+            if (typeList == null) {
+                return false;
+            }
+            for (SimpleOrderedMap type : typeList) {
+                String name = (String) type.get("name");
+                if (name.equals(typeName)) {
+                    return true;
+                }
+            }
+        } catch (Exception e) {
+            log.error(e, e);
+        }
+        return false;
+    }
+
+    private static boolean isDynamicFieldExists(String fieldName, SolrClient queryEngine,
+            ConcurrentUpdateSolrClient updateEngine) throws Exception {
+        try {
+            SolrQuery query = new SolrQuery();
+            query.add(CommonParams.QT, "/schema/dynamicfields");
+            QueryResponse response = queryEngine.query(query);
+            ArrayList<SimpleOrderedMap> fieldList = (ArrayList<SimpleOrderedMap>) response.getResponse()
+                    .get("dynamicFields");
+            if (fieldList == null) {
+                return false;
+            }
+            for (SimpleOrderedMap field : fieldList) {
+                String name = (String) field.get("name");
+                if (name.equals(fieldName)) {
+                    return true;
+                }
+            }
+        } catch (Exception e) {
+            log.error(e, e);
+        }
+        return false;
+    }
+
+    private static void createDateRangeFieldType(ConcurrentUpdateSolrClient updateEngine) throws Exception {
+        SchemaRequest.AddFieldType fieldTyperequest = new SchemaRequest.AddFieldType(getDateRangeFieldTypeDefinition());
+        SchemaResponse.UpdateResponse response = fieldTyperequest.process(updateEngine);
+        if (response.getStatus() != 0) {
+            throw new Exception("Creation of missing solr field type 'dateRange' failed");
+        }
+        log.info("Solr field type 'dateRange' has been created.");
+    }
+
+    private static void createDateRangeField(String suffix, ConcurrentUpdateSolrClient updateEngine) throws Exception {
+        Map<String, Object> fieldAttributes = getDateRangeFieldAttributes(suffix);
+        SchemaRequest.AddDynamicField dynamicFieldrequest = new SchemaRequest.AddDynamicField(fieldAttributes);
+        SchemaResponse.UpdateResponse response = dynamicFieldrequest.process(updateEngine);
+        if (response.getStatus() != 0) {
+            throw new Exception("Creation of missing solr field '*" + suffix + "' failed");
```

Comment:
I think that it is a better practice if we define new exception type for this scenario eg. SolrFieldCreationException, that way if we happen to implement a catch mechanism later down the line we would have an easier job differentiating between different exceptions thrown. You can make a new class that inherits `RuntimeException`, simple constructor with only `String message` should suffice.

Reply:
Fixed